File: surface/widgets/compose/components/widgets/domain/CodeBlock.kt
// ============================================================
// Clef Surface Compose Widget — CodeBlock
//
// Syntax-highlighted code display block rendered as a Surface
// with monospace Text and syntax highlighting colors. Shows
// optional line numbers, highlighted lines with a tinted
// background, and a copyable indicator in the header.
//
// Adapts the code-block.widget spec: anatomy (root, header,
// language, copyButton, lineNumbers, code), states (idle,
// hovered, focused, copied), and connect attributes.
// ============================================================

package clef.surface.compose.components.widgets.domain

import androidx.compose.foundation.background
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.delay

// --------------- Component ---------------

/**
 * Syntax-highlighted code display block with monospace text.
 *
 * @param code The source code string to display.
 * @param language Programming language name shown in the header.
 * @param showLineNumbers Whether to show line numbers in the gutter.
 * @param highlightLines Line numbers to highlight (1-based).
 * @param copyable Whether the code can be copied.
 * @param onCopy Callback invoked when the copy button is pressed.
 * @param modifier Modifier for the root Surface.
 */
@Composable
fun CodeBlock(
    code: String,
    language: String = "plaintext",
    showLineNumbers: Boolean = true,
    highlightLines: Set<Int> = emptySet(),
    copyable: Boolean = false,
    onCopy: (String) -> Unit = {},
    modifier: Modifier = Modifier,
) {
    var copied by remember { mutableStateOf(false) }

    LaunchedEffect(copied) {
        if (copied) {
            delay(2000)
            copied = false
        }
    }

    val lines = code.split("\n")
    val gutterWidth = if (showLineNumbers) lines.size.toString().length + 1 else 0

    Surface(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(8.dp),
        color = Color(0xFF1E1E1E),
        tonalElevation = 2.dp,
    ) {
        Column {
            // Header
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 12.dp, vertical = 8.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    text = language,
                    style = MaterialTheme.typography.labelMedium,
                    fontWeight = FontWeight.Bold,
                    color = Color(0xFFDCDCAA),
                )
                Spacer(modifier = Modifier.weight(1f))
                if (copyable) {
                    TextButton(onClick = {
                        onCopy(code)
                        copied = true
                    }) {
                        Text(
                            text = if (copied) "Copied!" else "Copy",
                            color = if (copied) Color(0xFF4EC9B0) else Color(0xFF9CDCFE),
                            style = MaterialTheme.typography.labelSmall,
                        )
                    }
                }
            }

            // Code lines
            Column(
                modifier = Modifier
                    .horizontalScroll(rememberScrollState())
                    .padding(horizontal = 12.dp, vertical = 4.dp),
            ) {
                lines.forEachIndexed { index, line ->
                    val lineNum = index + 1
                    val isHighlighted = lineNum in highlightLines

                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .then(
                                if (isHighlighted)
                                    Modifier.background(Color(0xFF264F78))
                                else
                                    Modifier,
                            ),
                    ) {
                        Row {
                            if (showLineNumbers) {
                                Text(
                                    text = lineNum.toString().padStart(gutterWidth),
                                    fontFamily = FontFamily.Monospace,
                                    fontSize = 13.sp,
                                    color = Color(0xFF858585),
                                )
                                Spacer(modifier = Modifier.width(12.dp))
                            }
                            Text(
                                text = line,
                                fontFamily = FontFamily.Monospace,
                                fontSize = 13.sp,
                                color = Color(0xFFD4D4D4),
                            )
                        }
                    }
                }
            }

            Spacer(modifier = Modifier.padding(bottom = 8.dp))
        }
    }
}

File: surface/widgets/swiftui/components/widgets/domain/CodeBlock.swift
// ============================================================
// Clef Surface SwiftUI Widget — CodeBlock
//
// Syntax-highlighted code display block rendered with monospace
// text. Shows optional line numbers, highlighted lines with a
// tinted background, and a copyable indicator in the header.
// ============================================================

import SwiftUI

struct CodeBlockView: View {
    var code: String
    var language: String = "plaintext"
    var showLineNumbers: Bool = true
    var highlightLines: Set<Int> = []
    var copyable: Bool = false
    var onCopy: (String) -> Void = { _ in }

    @State private var copied = false

    var body: some View {
        let lines = code.split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
        let gutterWidth = showLineNumbers ? max(String(lines.count).count + 1, 3) : 0

        VStack(alignment: .leading, spacing: 0) {
            // Header
            HStack {
                Text(language)
                    .font(.caption)
                    .fontWeight(.bold)
                    .foregroundColor(Color(red: 0.86, green: 0.86, blue: 0.67))

                Spacer()

                if copyable {
                    SwiftUI.Button(action: {
                        onCopy(code)
                        copied = true
                        DispatchQueue.main.asyncAfter(deadline: .now() + 2) {
                            copied = false
                        }
                    }) {
                        Text(copied ? "Copied!" : "Copy")
                            .font(.caption2)
                            .foregroundColor(copied ? Color(red: 0.31, green: 0.79, blue: 0.69) : Color(red: 0.61, green: 0.86, blue: 0.99))
                    }
                    .buttonStyle(.plain)
                }
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 8)

            // Code lines
            ScrollView(.horizontal, showsIndicators: false) {
                VStack(alignment: .leading, spacing: 0) {
                    ForEach(Array(lines.enumerated()), id: \.offset) { index, line in
                        let lineNum = index + 1
                        let isHighlighted = highlightLines.contains(lineNum)

                        HStack(spacing: 0) {
                            if showLineNumbers {
                                Text(String(lineNum).padding(toLength: gutterWidth, withPad: " ", startingAt: 0))
                                    .font(.system(size: 13, design: .monospaced))
                                    .foregroundColor(Color(white: 0.52))
                                    .frame(minWidth: CGFloat(gutterWidth * 8), alignment: .trailing)

                                Spacer().frame(width: 12)
                            }

                            Text(line)
                                .font(.system(size: 13, design: .monospaced))
                                .foregroundColor(Color(red: 0.83, green: 0.83, blue: 0.83))
                        }
                        .padding(.horizontal, 12)
                        .padding(.vertical, 1)
                        .background(isHighlighted ? Color(red: 0.15, green: 0.31, blue: 0.47) : Color.clear)
                    }
                }
                .padding(.vertical, 4)
            }

            Spacer().frame(height: 8)
        }
        .background(
            RoundedRectangle(cornerRadius: 8)
                .fill(Color(red: 0.12, green: 0.12, blue: 0.12))
        )
    }
}
